packages/common-all/src/store/IDataQuery.ts

```diff
@@ -0,0 +1,39 @@
+import { ResultAsync } from "neverthrow";
+import {
+  DendronError,
+  NoteChangeEntry,
+  NotePropsByIdDict,
+  NotePropsMeta,
+  SchemaModuleDict,
+  SchemaModuleProps,
+} from "..";
+
+export type INoteQueryOpts = {
+  /**
+   * Original query string (which can contain minor modifications such as mapping '/'->'.')
+   * This string is added for sorting the lookup results when there is exact match with
+   * original query. */
+  originalQS: string;
+  onlyDirectChildren?: boolean;
+};
+
+export interface IQueryStore {
+  queryNotes(
+    qs: string,
+    opts: INoteQueryOpts
+  ): ResultAsync<NotePropsMeta[], DendronError>;
+  querySchemas(
+    qs: string,
+    opts?: INoteQueryOpts
+  ): ResultAsync<{ id: string }[], DendronError>;
+  updateNotesIndex(changes: NoteChangeEntry[]): ResultAsync<void, DendronError>;
```

Comment:
IMO we shouldn't be exposing these `update`/`replace`/`remove` methods, because they're exposing too much implementation detail through the interface.  For example - why does an `IQueryStore` need to use indexes? This interface assumes it, which greatly reduces the utility of the interface.

I get that if we decide to not combine the Fuse logic with the metadata store, we won't be able to remove the current Fuse references from `EngineV3*` classes without adding some methods like these - but then at that point, we aren't _really_ removing Fuse details from those classes since all the assumptions on how the store works still exist through this interface.

As I was coding this up, I also realized that a queryStore implementation would be hard to decouple from the metadata store CRUD methods, so perhaps a more sensible thing to do here is to have IQueryStore extend `IDataStore<string, NotePropsMeta>`.  Not sure yet - dunno if it's more sensible that way vs having `IQueryStore` just expose `queryNotes`/`querySchema`. 

Unrelated - but can we make this file name align with the interface name as per our new convention? i.e. `IQueryStore.ts`

Reply:
i think all of these methods will still apply to any querystore. i don't want to create something generic until we have at least two implementations of it (aka - once we implement the SQLite interface, we will have a better idea of where to draw the lines). 

my thinking here is that this is the intermediary step. the interface is not *finalized* until we also implement the SQLIte store behind this interface.

ack on changing the name

Reply:
What is the benefit we are gaining from incrementally introducing a non-finalized interface when we only have one implementation of it? wouldn't it be better to wait for a second implementation to come, and then with the two extrapolate a generic interface instead of introducing more things to clean up in the future?

I can understand introducing some non-finalized methods in the interface if we are already dealing with more than one implementation of it. But I don't feel like we are gaining anything from incrementally refactoring at this point of time.

What is the timeline for implementing the SQLite store? If the gap between now and then is large enough, eventually these non-generics will be used in many places as if they were final. We've seen this in many places that we allowed intermediary non-finalized refactors (notably in config, though we could argue that it was necessary in config), and the cleanup was a heavy task. Given our total bandwidth, we are always hesitant on budgeting out the time to clean things up, which makes the matter even worse.

IMO, it would be okay if this PR was followed right away with the discussion for the SQLite store, and we immediately resolve the non-generic nature of it, but not if we don't have a reasonable ETA on when we are going to introduce the SQLite store.

Reply:
@hikchoi - agree with your concerns.  I'll have an async with a proposal out soon on how to proceed.